examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+      // interface entries, i.e., the entry for dof pair (j,i) is stored in
+      // interface_in(i,j).
+      for (unsigned int i = 0; i < copy_data.dofs_per_cell; ++i)
+        for (unsigned int j = 0; j < copy_data.dofs_per_cell; ++j)
+          if (mg_constrained_dofs.is_interface_matrix_entry(
+                copy_data.level,
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j]))
+            {
+              mg_interface_out[copy_data.level].add(
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j],
+                copy_data.cell_matrix(i, j));
+              mg_interface_in[copy_data.level].add(
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j],
+                copy_data.cell_matrix(j, i));
+            }
+    };
+
+    MeshWorker::mesh_loop(dof_handler.begin_mg(),
+                          dof_handler.end_mg(),
+                          cell_worker_mg,
+                          copier_mg,
+                          ScratchData<dim>(fe, quad_degree),
+                          CopyData(),
+                          MeshWorker::assemble_own_cells);
+  }
+
+
+  template <int dim>
+  std::unique_ptr<MGSmoother<Vector<double>>>
+  AdvectionProblem<dim>::create_smoother()
+  {
+    if (settings.smoother_type == "sor")
+      {
+        typedef PreconditionSOR<SparseMatrix<double>> Smoother;
+
+        auto smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 1.0 :
+                                                                       0.62));
+        smoother->set_steps(2);
+        return smoother;
+      }
+    else if (settings.smoother_type == "jacobi")
+      {
+        typedef PreconditionJacobi<SparseMatrix<double>> Smoother;
+        auto                                             smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 0.6667 :
+                                                                       0.47));
+        smoother->set_steps(4);
+        return smoother;
+      }
+    else if (settings.smoother_type == "block sor")
+      {
+        typedef RelaxationBlockSOR<SparseMatrix<double>, double, Vector<double>>
+          Smoother;
+
+        static MGLevelObject<typename Smoother::AdditionalData> smoother_data;
+        smoother_data.resize(0, triangulation.n_levels() - 1);
```

Comment:
Does the variable have to be `static`? It seems odd here.

Reply:
Yes, it appears they do. When you remove, you get an error complaining that "the object point to is no longer valid". I think it is because the object which uses smoother_data is a pointer.

Reply:
It would be nice to understand this further. `static` variables are generally difficult because, for example, they're not thread-safe. They lead to all sorts of problems if you later want to re-factor a code. So my preference would be if you could write up what specifically happens if it's not static, and then find a different solution.